usl_pipeline/usl_lib/usl_lib/transformers/feature_raster_transformers.py

```diff
@@ -0,0 +1,153 @@
+    lookup_np_func = numpy.vectorize(
+        lambda soil_class: (
+            0.0
+            if soil_class not in lookup_map
+            else property_getter(lookup_map[soil_class])
+        ),
+        cache=True,
+    )
+    return lookup_np_func(soil_classes_raster) / max_property_value
+
+
+def transform_to_feature_raster_layers(
+    elevation: geo_data.Elevation,
+    boundaries_polygons: Optional[list[Tuple[geometry.Polygon, int]]],
+    buildings_polygons: list[Tuple[geometry.Polygon, int]],
+    green_areas_polygons: list[Tuple[geometry.Polygon, int]],
+    soil_classes_polygons: list[Tuple[geometry.Polygon, int]],
+    soil_infiltration_configuration: geo_data.InfiltrationConfiguration,
+) -> npt.NDArray[numpy.float32]:
+    """Rasterize layers of feature matrix and sanitize the values.
+
+    Args:
+        elevation: The elevation data.
+        boundaries_polygons: List of tuples with boundaries polygons.
+        buildings_polygons: List of tuples with buildings polygons.
+        green_areas_polygons: List of tuples with green area polygons.
+        soil_classes_polygons: List of tuples with soil class polygons.
+        soil_infiltration_configuration: Soil infiltration configuration which is used
+            to look up infiltration properties for soil classes.
+
+    Returns:
+        3-dimensional matrix where 2 first dimensions are spacial (Y and X axes) and
+        third one combines 8 feature layers:
+         - elevation data,
+         - elevation mask,
+         - building footprint mask,
+         - green area mask,
+         - hydraulic conductivity,
+         - wetting front suction head,
+         - effective porosity,
+         - effective saturation.
+    """
+    raster_h = elevation.header.row_count
+    raster_w = elevation.header.col_count
+    nodata_value = elevation.header.nodata_value
+    if elevation.data is None:
+        raise ValueError("Elevation data missing")
+    elevation_raster = elevation.data.astype(dtype=numpy.float32)
+    elevation_mask_raster = numpy.zeros((raster_h, raster_w), dtype=int)
+    # setting mask raster to 1 for cells where elevation value is present (!=NODATA)
+    elevation_mask_raster[elevation_raster != nodata_value] = 1
+    empty_raster = numpy.zeros((raster_h, raster_w), dtype=numpy.float32)
+    # If boundaries data is not present we ignore any boundaries-related corrections.
+    if boundaries_polygons is not None:
+        # If boundaries data is present but empty it means that the whole matrix should
+        # be marked as NODATA.
+        boundaries_raster = (
+            empty_raster
+            if len(boundaries_polygons) == 0
+            else polygon_transformers.rasterize_polygons(
+                elevation.header, boundaries_polygons
+            )
+        )
+        # clearing elevation values and presence mask for cells outside boundaries
+        elevation_raster[boundaries_raster == 0] = numpy.float32(nodata_value)
```

Comment:
We don't want to pass -9999s forward for no data. That will mess up the scaling.
We should use 0 for no-data. The supplied `elevation_mask_raster` will give the model what it needs. Trying to use special values for no-data will confuse the model. Large values like -9999 are especially bad because of how they'll impact scaling.

Reply:
We discussed this offline and came to a conclusion that it doesn't matter which placeholder value is used where presence mask = 0 (this mask indicates NODATA cells) as long as it will be taken into account after scaling. So just for consistency I use the same value as used for other NODATA cells.